[SYCL][ESIMD] Add test for esimd copy constructor

vasilytric opened the pull request:
This PR provides test for `simd` copy constructor.
The test verifies that `simd`'s elements value is equal to values from reference data

Also was added `common.hpp` file that provide functions for constructing `reference data` and function for calling `simd` in a predefined context 

____________________________________________

This PR depend on PR https://github.com/intel/llvm-test-suite/pull/565
In PR#565 can be found these files can be found in this :
- `logger.hpp`
- `type_coverage.hpp`

vasilytric committed:
[SYCL][ESIMD] Add common file for test for simd constructors

This file provides functional that let obtain reference data and common
function for constructing simd object, that return's output data that
should be compared with reference data

vasilytric committed:
[SYCL][ESIMD] Add test for simd copy constructor

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+using allocator_type = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
+
+template <typename DataT>
+using vector_with_allocator = std::vector<DataT, allocator_type<DataT>>;
```
```suggestion
// Vector type to share data between device and host sides using USM allocator
template <typename DataT>
using shared_vector = std::vector<DataT, allocator_type<DataT>>;
```
Just as a suggestion

vasilytric replied:
I've done it in this commit: 573abdb97eeea16c695266a7f80c965d0ec8f989

yuriykoch replied:
Please, feel free to close the resolved ones. Looks like I cannot do it myself for some reason.

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// Common file for test on simd constructors.
```
```suggestion
/// Common file for tests on simd constructors.
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/c0932b4585590a88396f3bed2364b90da50d0afe#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R11)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+
+#include "../../../esimd_test_utils.hpp"
+#include "../logger.hpp"
+#include "../type_coverage.hpp"
```
```suggestion
#include "../common.hpp"
```
Let's track this change so we don't forget it

vasilytric replied:
I've done it in this commit: 460bbd9e31ee9f5944add34f5e6d68f71117b377

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+    } else if constexpr (std::is_same_v<DataT, sycl::half>) {
+      nan_val = 0;
+    }
+    return nan_val;
```
Let's add a `static_assert` for a non-floating point type at least. Also for `inf`

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L104-L105)
but `inf` could be obtained for any variable type

yuriykoch replied:
Is there any reason to have `0` as `infinity` available for signed and unsigned integral types? Looks like you are not using it for any non-floating type already..

vasilytric replied:
Yes, You are right. I've done it in this commit: b8bf6ac53b1678647d320ea502e92de424fbe894

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+    } else {
+      min_val = std::numeric_limits<DataT>::lowest();
+    }
+    return min_val;
```
```suggestion
    if constexpr (std::is_same_v<DataT, sycl::half>) {
      return 0;
    } else {
      return std::numeric_limits<DataT>::lowest();
    }
```

Here and for `max()`

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L94-L100) and in other functions

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+// Generate std::vector with reference according to current tested data type and
+// number of elements.
+template <typename DataT, int NumElems>
+void generate_ref_data(vector_with_allocator<DataT> &ref_data) {
```
```suggestion
vector_with_allocator<DataT> generate_ref_data() {
```
So it would be possible to optimize allocation using `std::vector::reserve()` or appropriate constructor call.

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L128-R121)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+  const DataT max_half = max / 2;
+  const DataT inf = value<DataT>::inf();
+  const DataT min_plus_one = min + 1;
+  const DataT max_minus_one = max - 1;
```
```suggestion
```
What's the intent of the `min_plus_one`, `min_minus_one`, `min_half`, `max_half`? What should we change in case we need to add `min +2` at some point?

vasilytric replied:
I've removed `max_minus_one ` and `min_plus_one` variables, instead of this use `max - 1` and `min  + 1` while fill the vector it [for example](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L142-R133)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+      ref_data.insert(ref_data.end(), {min, min_half, max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {min, min_plus_one, min_half, max,
+                                       max_minus_one, max_half, -1, 0});
```
Looks like in all cases the reference data for `NumElems == 1` is the subset of the reference data for the different branch.
As we are using `insert()` anyway, why don't make something like
```
ref_data.insert(ref_data.end(), /*set of values common for both cases*/);

if constexpr (NumElems != 1) {
    ref_data.insert(ref_data.end(), /*remaining set of values*/);
}
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L153-L160)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+      ref_data.insert(ref_data.end(), {min, min_plus_one, min_half, max,
+                                       max_minus_one, max_half, -1, 0});
+      for (size_t it = 8; it < NumElems; it++) {
+        ref_data.push_back(it);
```
Minor:
(1) usually it's expected to see `it` meaning `iterator` in loops
(2) be aware it's generally better to prefer [pre-increment](https://llvm.org/docs/CodingStandards.html#prefer-preincrement) because it doesn't create an extra temporary value

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L147-L149)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+  const DataT min_half = min / 2;
+  const DataT max = value<DataT>::max();
+  const DataT max_half = max / 2;
+  const DataT inf = value<DataT>::inf();
```
Let's move it to the branch for the floating point types

vasilytric replied:
I've moved `inf` o the branch for the floating point types [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R152)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+      ref_data.insert(ref_data.end(), {max, max_half, 0});
+    } else {
+      ref_data.insert(ref_data.end(), {max, max_minus_one, max_half, 0});
+      for (size_t it = 4; it < NumElems; it++) {
```
```suggestion
      for (size_t i = ref_data.size(); i < NumElems; ++i) {
```
Avoiding magic constants may make our code more robust - so, for example, we might insert() more than `4` values and do not need to change this loop.

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R134)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+  }
+
+  else if constexpr (std::is_same_v<DataT, sycl::half> ||
+                     std::is_floating_point_v<DataT>) {
```
```suggestion
  else if constexpr (type_traits::is_floating_point_v<DataT>) {
```
We might need it sooner or later.. May be it's out of scope for this PR though

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R150)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+
+  else if constexpr (std::is_same_v<DataT, sycl::half> ||
+                     std::is_floating_point_v<DataT>) {
+    DataT nan{value<DataT>::nan()};
```
```suggestion
    static const DataT nan{value<DataT>::nan()};
```
The idea behind the changes is as follows:
(1) const-correctness guarantees we don't change this value later, so it's more like an alias than like a variable
(2) static within the function scope enables the lazy-initialization - with no need to re-initialize with every call of `generate_ref_data`
I would suggest to apply such changes for all `value<DataT>::` synonyms in this function

vasilytric replied:
I've done `static_assert` in this commit: 74c17944c3316f44492338a96ffa9d2ec008a0c5

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+      ref_data.insert(ref_data.end(),
+                      {-inf, nan, min, max, -0.1, -0.0, +0.0, +0.1});
+      for (size_t it = 8; it < NumElems; it++) {
+        ref_data.push_back(it + 0.25);
```
```suggestion
        // Store values with exact representation of the fraction part for
        // every floating point type
        ref_data.push_back(it + 0.25);
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3c4a51f2dd397e3fb56059d085f481041e558a26#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R158-R159)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+        ref_data.push_back(it + 0.25);
+      }
+    } // else
+  }
```
I would suggest to

- replace `if - else if` logic with simple `if constexpr` branches
- add `static_assert` at the beginning of function to (1) validate the `DataT` is as expected (2) document DataT preconditions

vasilytric replied:
I've removed `if else` [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514L153) and insert `if` [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3f3cc72062810675bd658f89ec1e2ce1478da3a1#diff-f085fe3f94e6fdb4bc2da0518bd492beb702cd3f8181a461be78bcc561ccc514R140)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+  vector_with_allocator<DataT> result{NumElems, allocator_type<DataT>{queue}};
+
+  queue.submit([&](sycl::handler &cgh) {
+    auto ref = ref_data.data();
```
```suggestion
    const auto ref = ref_data.data();
```
Is it possible to enforce const-correctness for the reference data?

vasilytric replied:
I've done it in this commit: 0bab576359c48f2c44e6362dd3a7cb18d517def8

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+// Calling simd constructor in provided invocation context, it depends on the
+// TestCaseT type. TestCaseT is a struct, that should have call_simd_ctor method
+// that should return constructed object of simd class.
+// This function returning std::vector with output data.
```
```suggestion
// This function returns std::vector instance with the output data.
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3c4a51f2dd397e3fb56059d085f481041e558a26#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R44)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/common.hpp:
```diff
@@ -0,0 +1,165 @@
+  }
+}
+
+// Calling simd constructor in provided invocation context, it depends on the
```
```suggestion
// Calls simd constructor in provided invocation context, which depends on the
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3c4a51f2dd397e3fb56059d085f481041e558a26#diff-ef0766a5c874c909fd25e4eec71893c4614c91112485144e4efcdab4d856d0a0R41)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+
+#include "common.hpp"
+
+using namespace cl::sycl;
```
Actually.. Is there anything that stops us from using `<sycl/sycl.hpp>` with `::sycl` namespace? Looks like it would be better to switch to the SYCL-2020 headers from the start.

vasilytric replied:
I've done it in this commit: d5856d2f986efc316550d97a55a928e952a9417f

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+using namespace esimd_test::api::functional;
+
+// Struct that provide method that let get current test case description and
+// method that let call simd constructor for initializer test case
```
```suggestion
// Descriptor class for the case of calling constructor in initializer context
```
As a generic hint - let's describe more "why" instead of "what". And let's avoid duplicating the code logic in the comments.
By doing so, we:

- provide laconic hints for intent (which are required for review and later code updates)
- lesser the possibility we would need to fix the comments - for example, if we add the 3rd method

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3c4a51f2dd397e3fb56059d085f481041e558a26#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581R21)

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3c4a51f2dd397e3fb56059d085f481041e558a26#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581R21)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems> call_simd_ctor(const DataT *ref_data) {
+    simd<DataT, NumElems> source_simd{};
```
```suggestion
    simd<DataT, NumElems> source_simd;
```
Let's enforce default-construction. Take a look [here](https://llvm.org/docs/CodingStandards.html#do-not-use-braced-initializer-lists-to-call-a-constructor)

This comment can be applied for any test case descriptor.

vasilytric replied:
I've done it in this commit: 8710841ed7c8749ad9e606a45045c7ffc0d7dbae

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+
+  template <typename DataT, int NumElems>
+  static simd<DataT, NumElems>
+  call_simd_by_const_ref(const simd<DataT, NumElems> &simd_by_const_ref) {
```
Looks like we need a `call_simd_by_const_ref` to be private. May be we may use `class` for such case?

By doing so we make it explicit for the fellow programmers that the function is not a part of the public API.

vasilytric replied:
I've done it in this commit: f3b1c77af045eb36a185fde9525b8b5b3bd2d733

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+};
+
+// The main test routine.
+// Use struct with overloaded call operator to have able for iterating over
+// pre-determined datatypes.
```
```suggestion
// Using functor class to be able to iterate over the pre-defined data types
```

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/3c4a51f2dd397e3fb56059d085f481041e558a26#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581R84)

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+    vector_with_allocator<DataT> ref_data{allocator_type<DataT>{queue}};
+    generate_ref_data<DataT, NumElems>(ref_data);
+
+    auto result_data = call_simd<DataT, NumElems, TestCaseT>(queue, ref_data);
```
```suggestion
    const auto result_data = call_simd<DataT, NumElems, TestCaseT>(queue, ref_data);
```

vasilytric replied:
I've done it in this commit: 92b81b151170f0570d3d71b4ec9c78c211298b9c

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+using run_test_with_thirty_two_elems = test<DataT, 32, TestT>;
+
+template <typename TestT, typename... T>
+bool run_verification_with_chosen_test_type(
```
What about something like `for_all_types_and_dims(named_type_pack, integer_pack)` so we don't need all these wrappers?
We might strictly need it for other tests - still looks like out of scope of this PR though)

vasilytric replied:
I can done it while I will implementing tests for fill constructor

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp:
```diff
@@ -0,0 +1,153 @@
+  auto types{get_tested_types<tested_types::all>()};
+
+  passed &= run_verification_with_chosen_test_type<initializer>(queue, types);
+  passed &=
+      run_verification_with_chosen_test_type<var_declaration>(queue, types);
```
```suggestion
  ok &= run_verification_with_chosen_test_type<var_declaration>(queue, types);
```
or
```suggestion
  passed &= run_verification_for_type<var_declaration>(queue, types);
```
As a minor suggestion: by shorter names we are able to fit in a single line, so that `var_declaration` parameter is clearly on the same indent level with any other line

vasilytric replied:
I've done it [here](https://github.com/intel/llvm-test-suite/pull/578/commits/ec90c5248212600055572f138d681b54f6ab78c0#diff-e20a89556dfa9ec98f8eec4bea698de3975b780ff24a3bde1ced890af1d16581R143-R146)

vasilytric committed:
[SYCL][ESIMD] Move common.hpp into SYCL\ESIMD\api\functional

This file will be useful for another tests on simd

vasilytric committed:
[SYCL][ESIMD] Updating logic for comparing and initializing variables

For this test required init fp variables by bit statements and comparing
fp variables by bitwise comparison.

Co-Authored-By: Kochetkov, Yuriy <YuriyX.Kochetkov@intel.com>

vasilytric committed:
[SYCL][ESIMD] Update logic for generating reference data

- returning values without assignment it to variable
- returning vector with ref data instead of interacting with vector,
  that obtained by reference
- replacing if else block by if block because we doesn't provide some
  "else" block
- remove duplicated values for init ref values in vector
- rename "it" to "i" in sycles due to it's more logical

vasilytric committed:
[SYCL][ESIMD] Update comments

Update comments in the following files:
- SYCL/ESIMD/api/functional/common.hpp
- SYCL/ESIMD/api/functional/ctors/ctor_copy.cpp
- SYCL/ESIMD/api/functional/value.hpp

vasilytric committed:
[SYCL][ESIMD] Update formatting in ctor_copy.cpp

Previous function and return value assignment was in a different lines,
now they containings in the one line.

vasilytric committed:
[SYCL][ESIMD] Replace including <CL/sycl.hpp> to <sycl/sycl.hpp>

It would be better to switch to the SYCL-2020 headers

vasilytric committed:
[SYCL][ESIMD] Rename vector_with_allocator and shared_allocator

Renamed to shared_vector and shared_allocator.
This names is better than privious names.

vasilytric committed:
[SYCL][ESIMD] Make reference data const

To enforce const-correctness for the reference data

vasilytric committed:
[SYCL][ESIMD] Enforce default-construction in copy constructor test

LLVM Coding Standards required this.

vasilytric committed:
[SYCL][ESIMD] Make const_ref struct a class type

By doing so we make it explicit for the fellow programmers that
the function is not a part of the public API.

vasilytric committed:
[SYCL][ESIMD] Make result_data constant variable

It't more correctness

vasilytric committed:
[SYCL][ESIMD] Add static asserd for generate_ref_data function

It protect newly programmers for some unexpected errors

vasilytric committed:
[SYCL][ESIMD] Add static_assert(is_fp_type<T>) for inf() function

This function return zero for signed and unsigned datatypes, so this
function will be called only for floating poin datatypes.

vasilytric committed:
[SYCL][ESIMD] Add reserving storage in vector with reference data

This allows increase runtime performance

vasilytric committed:
[SYCL][ESIMD] Add static_cast for non const values in initializer list

This allows avoid using variables for expression like "max - 1"

vasilytric committed:
[SYCL][ESIMD] Move common functions and some methods to common files

It let avoid copy-paste of some code

vasilytric committed:
[SYCL][ESIMD] Add namespaces to common files in functional\ctors folder

It let avoid some double definition errors

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -0,0 +1,92 @@
+using shared_vector = std::vector<DataT, shared_allocator<DataT>>;
+
+// Calls simd constructor in provided invocation context, which depends on the
+// TestCaseT type. TestCaseT is a struct, that should have call_simd_ctor method
```
Is this comment relevant for the `functional/common.hpp`?
Moreover, I'm not quite sure we will need the same `call_simd` logic for every group of tests. For example, we might need two output buffers. This function looked on it's place while being in `functional/ctors/common.hpp` as the one specific to the constructor tests.

vasilytric replied:
this comment was applied in this commit: 3e7019d1753cffdbe4a35aafaa674d765639be0e

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -0,0 +1,92 @@
+template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
+  constexpr size_t size{sizeof(T)};
+
+  const auto &lhs_bytes = reinterpret_cast<const unsigned char(&)[size]>(lhs);
```
```suggestion
  // Such type-punning is OK from the point of strict aliasing rules
  const auto &lhs_bytes = reinterpret_cast<const unsigned char(&)[size]>(lhs);
```

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -0,0 +1,92 @@
+  return result;
+}
+
+// Function that comparing two provided variables. If this variable is floating
+// point type or signed type this variables will be bitwise compared, othervise
+// variables will be compared by "==" operator.
+template <typename T> bool are_equal(T lhs, T rhs) {
+  if constexpr (type_traits::is_floating_point_v<T> || std::is_signed_v<T>) {
+    return are_bitwise_equal(lhs, rhs);
```
```suggestion
// A wrapper to speed-up bitwise comparison
template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
  if constexpr (type_traits::is_floating_point_v<T> || std::is_signed_v<T>) {
    return details::are_bitwise_equal(lhs, rhs);
```
or may be we don't need a separate wrapper at all?

General comment - please avoid duplicating code logic in comments.

vasilytric replied:
I tell You about this wrapper, in the test plan **all** types should be bitwise compared 
What do You think about this function? should we delete it and bitwise comparing all data types or this implementation is ok?
The comment for this function will be updated in one of next commits

vasilytric replied:
this comment was applied in this commit: 848b25756a3ca77feefbb306b2b620e384439b46
also use update can be found here: 1f1f8b9d296773e2495ab9330c86dad1f180df93

yuriykoch commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -0,0 +1,92 @@
+  return result;
+}
+
+// Bitwise comparing two provided variables.
```
```suggestion
// Bitwise comparison for two values
```

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/base_structs_with_invocation_ctx.hpp:
```diff
@@ -0,0 +1,59 @@
+// Descriptor class for the case of calling constructor in initializer context.
+struct initializer_base {
+  static std::string get_description() { return "initializer"; }
+};
```
Not quire sure we need a separate class for this.
Look, it's harder to read the code once `initializer`, `initializer_base` and `call_simd` are moved to the separate files.

vasilytric replied:
this comment was applied in this commit: 3e7019d1753cffdbe4a35aafaa674d765639be0e

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/base_structs_with_invocation_ctx.hpp:
```diff
@@ -0,0 +1,59 @@
+  call_simd_by_const_ref(
+      const sycl::ext::intel::experimental::esimd::simd<DataT, NumElems>
+          &simd_by_const_ref) {
+    return simd_by_const_ref;
```
I assume it wouldn't compile.
Yes, we cannot use polymorphism on the device side.
And yes, we might use "curiously recurring template pattern" to accomplish the static inheritance.

But does it worth it? Once again, it's too small and placed over multiple files, so it would be hard for me to get the idea if I haven't seen the version before. Generally, that's the level when it starts being hard to read an support.

vasilytric replied:
this comment was applied in this commit: 3e7019d1753cffdbe4a35aafaa674d765639be0e

yuriykoch commented on SYCL/ESIMD/api/functional/ctors/esimd_ctors_common.hpp:
```diff
@@ -0,0 +1,55 @@
+//===-- esimd_ctors_common.hpp - Define common code for simd ctors tests --===//
```
```suggestion
//===-- common.hpp - Define common code for simd ctors tests --===//
```
See, you already have "esimd" and "ctors" in the path: `ESIMD/.../ctors/common.hpp`, so why duplicate?

vasilytric replied:
this comment was applied in this commit: 3e7019d1753cffdbe4a35aafaa674d765639be0e

yuriykoch commented on SYCL/ESIMD/api/functional/type_traits.hpp:
```diff
@@ -0,0 +1,38 @@
+template <typename T> struct is_floating_point {
+  static constexpr bool value{std::is_floating_point_v<T> ||
+                              std::is_same_v<T, sycl::half>};
+};
```
```suggestion
template <typename T>
using is_sycl_floating_point =
  std::bool_constant<std::is_floating_point_v<T> ||
                                  std::is_same_v<T, sycl::half>>;

```
What do you think? It's available in C++17
1) At the closer look we have name clash wit std::is_floating_point here.
2) By using type alias we avoid creating any new type

But we will need to change the `is_sycl_type()` naming appropriately in case of such change. May be `is_sycl_specific<T>()`?

vasilytric replied:
this comment was applied in this commit: ef6509579513a2452045f0c2cf2c0599a6c3b96b

yuriykoch replied:
So you've preferred to retain the `is_sycl_type()` naming?
Actually.. Do we even use it? May be we should just remove the `is_sycl_type()` function?

vasilytric replied:
> May be we should just remove the is_sycl_type() function?

Yes, this function was removed in this commit: this function was removed in this commit: 465eb5fe27f50118853554671143841f92d49407

yuriykoch commented on SYCL/ESIMD/api/functional/type_traits.hpp:
```diff
@@ -0,0 +1,38 @@
+};
+
+template <typename T>
+inline constexpr bool is_floating_point_v{is_floating_point<T>::value};
```
```suggestion
inline constexpr bool is_sycl_floating_point_v{is_floating_point<T>::value};
```
In case of `is_sycl_floating_point` of course)

vasilytric replied:
this comment was applied in this commit: ef6509579513a2452045f0c2cf2c0599a6c3b96b

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+  To dst;
+  std::memcpy(&dst, &src, sizeof(To));
+  return dst;
+}
```
```suggestion
```
Let's replace it with a single include of the `esimd_test_utils.hpp`

vasilytric replied:
this comment was applied in this commit: 9b470835fbdf8fd1c8ef174e34b26d62710a13cb

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+//  - the byte order of both the unsigned types and the floating type itself;
+//  - the existence of the optional std::uint16_t type
+//  - compiler optimisations related to the strict aliasing rules
+sycl::half half_from_bytes(unsigned char hi, unsigned char lo) noexcept {
```
Do we need to access it outside of this header?
Let's move to the `functional::details` namespace within this header if not.

And actually sorry for the `noexcept` guarantee. Let's remove it, otherwise application will abort on every exception within such function.

vasilytric replied:
this comment was applied in this commit: 3e7019d1753cffdbe4a35aafaa674d765639be0e

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+  unsigned char index[size];
+  unsigned char out[size];
+
+  sycl::half indexHint = 2;
+  const unsigned char indexCoeff = 64;
```
```suggestion
  // We are using specific half value to initialize the bits required to differ
  // the lowest and the highest byte
  sycl::half indexHint = 2;
  const unsigned char indexCoeff = 64;
```

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+  index[0] /= indexCoeff;
+  index[1] /= indexCoeff;
+
+  assert(index[0] + index[1] == 1);
```
```suggestion
  // Ensure there is no overflow possible
  assert(index[0] + index[1] == 1);
```

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+template <typename DataT> struct value {
+  static DataT inf() {
+    static_assert(type_traits::is_floating_point_v<DataT>,
+                  "Infinity required only for floating point data types.");
```
```suggestion
                  "Infinity is required only for the floating point data types.");
```

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+}
+
+// Struct that provise methods for constructing inf, lowest, nan, denorm_min and
+// max values for reference data.
```
```suggestion
// Utility class to retrieve specific values for tests depending on the data type
// May be used to retrieve reference data or for generation of golden values
```

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+
+  static DataT lowest() {
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return -half_from_bytes(0b01111011u, 0b11111111u);
```
Either
```suggestion
      return -max();
```
or
```suggestion
      return half_from_bytes(0b1111011u, 0b11111111u);
```
The first one is the preferred one - because we don't duplicate the actual bits.

vasilytric replied:
this comment was applied in this commit: 4bd389b51699b9b2e1d891e8bcab1c536cd45a90

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+  static DataT nan(unsigned char opcode = 42u) {
+    static_assert(
+        type_traits::is_floating_point_v<DataT>,
+        "nan can value can retrieved only for floating point variable type.");
```
Let's have the similar description as for infinity.

```suggestion
        "NaN has meaning only for floating point data types.");
```

By making descriptions uniform we are making differences more explicit

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+    if constexpr (std::is_same_v<DataT, double>) {
+      return sycl::nan(42UL);
+    } else if constexpr (std::is_same_v<DataT, float>) {
+      return sycl::nan(42U);
```
Whoops..
Either use `opcode`  everywhere, or use the `42` explicitly. Why should this function ignore parameter for any type except the half?

If you are going to use parameter - don't forget about static_cast for the `unsigned long long` branch.

vasilytric replied:
this comment was applied in this commit: 4bd389b51699b9b2e1d891e8bcab1c536cd45a90

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+  }
+};
+
+// Filling obtained std::vector with reference data according to current tested
```
```suggestion
// Provides std::vector with the reference data according to the currently tested
```
Just as a suggestion )

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+// data type and number of elements.
+template <typename DataT, int NumElems> std::vector<DataT> generate_ref_data() {
+  static_assert(
+      std::is_integral_v<DataT> || std::is_arithmetic_v<DataT> ||
```
```suggestion
      std::is_integral_v<DataT> ||
```
Arithmetic type contains a much bigger subset than we are going to cover. Do we really need it?
Or did you mean `arithmetic && (signed || floating)` may be?

vasilytric replied:
this comment was applied in this commit: 4bd389b51699b9b2e1d891e8bcab1c536cd45a90

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+
+    ref_data.insert(ref_data.end(), {min, min_half, max, max_half, 0});
+    if constexpr (NumElems != 1) {
+      ref_data.insert(ref_data.end(), {static_cast<DataT>(min + 1),
```
Why do you need the static_cast calls here?
Here and elsewhere with `insert()` calls within this function

vasilytric replied:
this comment was applied in this commit: 4bd389b51699b9b2e1d891e8bcab1c536cd45a90
now I use `min_plus_one` and `max_minus_one` variables, that I was used earlier

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+
+    ref_data.reserve((NumElems > 1) ? NumElems : 6);
+
+    ref_data.insert(ref_data.end(), {-inf, nan, min, max, -0.0, 0.1});
```
```suggestion
    // We are using the `double` literals to avoid precision loss for case of
    // the `double` DataT on unexact values like 0.1
    ref_data.insert(ref_data.end(), {-inf, nan, min, max, -0.0, 0.1});
```

vasilytric replied:
this comment was applied in this commit: a1b88dbf574ff747afa3a0546423fdc6ce87a587

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,185 @@
+
+  if constexpr (type_traits::is_floating_point_v<DataT>) {
+    static const DataT nan{value<DataT>::nan()};
+    static const DataT inf = value<DataT>::inf();
```
```suggestion
    static const DataT nan = value<DataT>::nan();
    static const DataT inf = value<DataT>::inf();
```
Please be consistent over the code if it possible. Avoid using the different ways for the same thing.

vasilytric replied:
this comment was applied in this commit: 4bd389b51699b9b2e1d891e8bcab1c536cd45a90

yuriykoch reviewed (commented):
Thanks, Vasily

vasilytric committed:
[SYCL][ESIMD] Update comments for common.hpp and value.hpp

vasilytric committed:
[SYCL][ESIMD] Structures changes in test for copy constructor

- Uses do not need see some file-specific functions, so they moved into
"details" namespace
- Base clases was removed due to they could confuse new users
- Ctors specific functions were moved to ctors/common.hpp

vasilytric committed:
[SYCL][ESIMD] Small changes in value.hpp

- remove static_case and use static variables min_plus_one and
  max_minus_one due to this better solutions
- remove uniform initializations
- remove std::is_arithmetic_v due to by this condition can pass data
  types that shouldn't be passed to the generate_ref_data function

vasilytric committed:
[SYCL][ESIMD] Remove bit_cast due to it defined in esimd_test_utils.hpp

vasilytric committed:
[SYCL][ESIMD] Rename "are_equal" to "are_bitwise_equal"

Due to the fact this name more self-descriebe

vasilytric committed:
[SYCL][ESIMD] Rename is_floating_point_v to is_sycl_floating_point_v

vasilytric committed:
[SYCL][ESIMD] Run bitwise comparison if provided tye is unsigned

This let us comparing user-defined data types

yuriykoch commented on SYCL/ESIMD/api/functional/type_traits.hpp:
```diff
@@ -0,0 +1,39 @@
+template <typename T> constexpr bool is_sycl_type() {
+  return std::is_same_v<T, sycl::half>;
+}
+
```
```suggestion
```
Looks like we don't use it anyway :)

vasilytric replied:
this function was removed in this commit: 465eb5fe27f50118853554671143841f92d49407

yuriykoch commented on SYCL/ESIMD/api/functional/value.hpp:
```diff
@@ -0,0 +1,183 @@
+      std::is_integral_v<DataT> || type_traits::is_sycl_floating_point_v<DataT>,
+      "Invalid data type provided to the generate_ref_data function.");
+
+  static const DataT min = value<DataT>::lowest();
```
```suggestion
  // Create values with the strict type guarantee
  static const DataT min = value<DataT>::lowest();
```
Just as a suggestion - we may add a comment to avoid confusion about `min_plus_one` and `max_minus_one` values

vasilytric replied:
comment was added in this commit: 55ea733ea6bcac291953e9b4bf66a23c4ed7f275

yuriykoch reviewed (commented):
Thanks Vasily) The code looks great.
Looks like it depends on PR #565 - please mark the dependency in description explicitly.

vasilytric committed:
[SYCL][ESIMD] Add comment for values in generate_ref_data function

vasilytric committed:
[SYCL][ESIMD] Remove is_sycl_type because it dont use anywhere

vasilytric commented:
> Looks like it depends on PR #565 - please mark the dependency in description explicitly.

Thanks for the advice, done

v-klochkov commented on SYCL/ESIMD/api/functional/common.hpp:
```diff
@@ -0,0 +1,62 @@
+} // namespace details
+
+// A wrapper to speed-up bitwise comparison
+template <typename T> bool are_bitwise_equal(T lhs, T rhs) {
```
Sorry, I don't understand this code. Is it needed to speed-up all cases except unsigned integral types by  using == for them? Why not to use regular == for unsigned types?
It may be Ok to use that slower detail::are_bitwise_equal() method for types that do not have operator== defined, but for all other types including unsigned types == is good.

yuriykoch replied:
Right.. Looks like a misprint, the original intent was to limit the bitwise comparison for:

- signed integer type - due to 3 representations available
- floating point type - due to nan values with opcode

@vasilytric ?

v-klochkov replied:
I understand why detail::are_bitwise_equal() may be needed for float (because of nan, etc), but why can't you use regular operation == for signed integer types?

yuriykoch replied:
What if there is a trap representation (negative zero) stored within the signed integer variable due to some test/product issue?
Not quite sure, but isn't it an UB to access it - so we may not expect any deterministic results from the test, incl. failure on different signed integer values?

vasilytric replied:
@v-klochkov was added comment, that explains why we cant compare all data type by `== operation`, also now `== operation` will be used for unsigned data type for all other will be used bitwise comparison
This change can be found in this commit: ddf461a2a5f4b538f7c331f2ca8b0a9970b18647

v-klochkov replied:
"What if there is a trap representation (negative zero) stored within the signed integer variable"
Hmm, what do you mean by "negative zero for signed integer". Signed integer does not hold negative zero. Negative zero can be stored in FP value, but not in integer.
I may not understand some detail of implementation, please explain.

yuriykoch replied:
Actually I was quite surprised as well. It's well-known that signed integral values overflow is UB. But the details are much more interesting.

SYCL 2020 works with C++17. Prior to C++20 there were 3 possible binary representations for signed integral types (see [basic.fundamental] in C++17):

- two's complement
- ones' complement
- signed magnitude

From the list above only the `two's complement` representation doesn't have a "negative zero" value, so every other representation has AFAIK a so-called "trap representation":

- ones' complement: all bits set to `1`
- signed magnitude: first bit set to `1`, the rest of bits set to `0`

One of the ways to retrieve such values is to have `INT_MAX +1` or `INT_MIN - 1`, but there is definitely much more ways to do so - for example, by using bitwise operators.

SYCL 2020 guarantees the binary representation on host side and device side are the same. But it doesn't have any additional guarantees on actual binary representation for the signed integral types.

C++17 is based on C11, which provides the definition of "trap representation" in p.6.2.6.1 (representation of types, general). Basically any variable may be initialized to store such representation without causing UB, but the value of the variable cannot be used until a proper value stored in it.

v-klochkov replied:
Thank you for explanations.

vasilytric committed:
[SYCL][ESIMD] Fix in are_bitwise_equal function

- Unsigned variables will be compared by == operation all other types will
be compared bitwise
- Added comment that explaining need to use bitwise comparison

vasilytric committed:
Merge branch 'intel' into simd_ctor_coy

vasilytric committed:
[SYCL][ESIMD] Add "XREQUIRES" field and description of unsupp platform

We requiring gpu and level_zero platforms due to the tests on simd
constructors doesn't supported OPenCL platform yet

vasilytric committed:
[SYCL][ESIMD] Add TEST_WITH_COPY_FROM macro into ctor_copy tests

I've got unexpected static assertion when calling simd::copy_from()
function. Was created issue in intel/llvm and this macros must be
enabled when it is resolved

vasilytric committed:
[SYCL][ESIMD] Replace "RefType" with "DataT" in functional/value.hpp

vasilytric committed:
[SYCL][ESIMD] Disable ctor_copy test for esimd API

The test disabled because of the fact the static_assert was retrieved
while calling simd::copy_from() function

vasilytric commented:
@v-klochkov  The test for `simd` default constructor was merged into this PR, all checks are passed
Could you review new commits? If all is ok could you approve and merge it?
